clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -459,10 +458,12 @@ private static int parseAcks(String acksString) {
             int partition = partition(record, serializedKey, serializedValue, metadata.fetch());
             int serializedSize = Records.LOG_OVERHEAD + Record.recordSize(serializedKey, serializedValue);
             ensureValidRecordSize(serializedSize);
-            TopicPartition tp = new TopicPartition(record.topic(), partition);
+            tp = new TopicPartition(record.topic(), partition);
             long timestamp = record.timestamp() == null ? time.milliseconds() : record.timestamp();
             log.trace("Sending record {} with callback {} to topic {} partition {}", record, callback, record.topic(), partition);
-            RecordAccumulator.RecordAppendResult result = accumulator.append(tp, timestamp, serializedKey, serializedValue, callback, remainingWaitMs);
+            // producer callback will make sure to call both 'callback' and interceptor callback
+            Callback interceptCallback = this.interceptors == null ? callback : new InterceptorCallback<>(callback, this.interceptors, tp);
```

Comment:
Just thinking it loud here. Would it make sense to always combine user callbacks and interceptor callbacks, irrespective of whether interceptors are null or not. I think that can help in avoiding `this.interceptors != null` check at bunch of places.